prow/sidecar/run.go

```diff
@@ -121,13 +122,29 @@ func (o Options) Run() error {
 		aborted = returnCode == 130
 	}
 
-	return o.doUpload(spec, passed, aborted)
+	metadata := map[string]string{}
+	metadataRaw := o.GcsOptions.GetNamedArtifactsContent("metadata.json")
+	if len(metadataRaw) > 0 {
+		if err := json.Unmarshal(metadataRaw, &metadata); err != nil {
+			logrus.WithError(err).Warn("Failed to unmarshal metadata.json")
+		}
+	}
+
+	return o.doUpload(spec, passed, aborted, metadata)
 }
 
-func (o Options) doUpload(spec *downwardapi.JobSpec, passed, aborted bool) error {
+func (o Options) doUpload(spec *downwardapi.JobSpec, passed, aborted bool, metadata map[string]string) error {
 	uploadTargets := map[string]gcs.UploadFunc{
 		"build-log.txt": gcs.FileUpload(o.WrapperOptions.ProcessLog),
 	}
+
+	specData, err := yaml.Marshal(spec)
+	if err != nil {
+		logrus.WithError(err).Warn("Could not marshal spec data")
+	} else {
+		uploadTargets["spec.yaml"] = gcs.DataUpload(bytes.NewBuffer(specData))
+	}
```

Comment:
probably should use full prowjob spec instead?

Reply:
Yeah that would be more useful, but I don't think we pass that down to the job currently. Maybe we should be exposing the ProwJob spec and the Pod spec to jobs somehow? I think for now the jobspec could be sufficient, but we should probably choose a different filename to distinguish job spec from prowjob spec and pod spec in case we add those later.
@stevekuznetsov  WDYT?

Reply:
I think we need to do a breakout ... the original thought was that there would just be a controller that does this since all the data on where to put  stuff is in the PJ CRD. The controller exists (`artifact-uploader`) and just needs to be fixed up a little bit.

Reply:
let's chat in sig-testing?

Reply:
gonna drop the spec upload for now until we have a more proper implementation, per discussion from sig-testing